jib-core/src/main/java/com/google/cloud/tools/jib/ncache/CacheStorage.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.ncache;
+
+import com.google.cloud.tools.jib.image.DescriptorDigest;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Interface for queries to a cache storage engine.
+ *
+ * <p>The cache storage engine stores layer data as {@link CacheWriteEntry}s. These entries are read
+ * out as {@link CacheReadEntry}s. Cache entries can be retrieved by the layer digest.
+ *
+ * <p>The cache entries can also be queried by an arbitrarily-defined selector (in digest format).
+ * The selectors do not need to be unique. An example of a selector could be the digest of the list
+ * of source file paths that constructed the cached layer.
+ *
+ * <p><b>Implementations must be thread-safe and should be immutable.</b>
+ */
+public interface CacheStorage {
+
+  /**
+   * Saves the {@link CacheWriteEntry}.
+   *
+   * @param cacheWriteEntry the {@link CacheWriteEntry}
+   * @return the {@link CacheReadEntry} for the written {@link CacheWriteEntry}
+   * @throws IOException if an I/O exception occurs
+   */
+  CacheReadEntry save(CacheWriteEntry cacheWriteEntry) throws IOException;
```

Comment:
I don't know how to create a CacheWriteEntry.  Why not have this method instead take the required pieces:
```
CacheReadEntry saveLayer(Blob layerBlob, @Nullable Blob metadataBlob, @Nullable DescriptorDigest selector);
```